algorithms/HPdclassifier/R/hpdsampling.R

```diff
@@ -0,0 +1,115 @@
+# of the License, or (at your option) any later version.                            
+#                                                                                   
+# This program is distributed in the hope that it will be useful,                   
+# but WITHOUT ANY WARRANTY; without even the implied warranty of                    
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                      
+# GNU General Public License for more details.                                      
+#                                                                                   
+# You should have received a copy of the GNU General Public License                 
+# along with this program; if not, write to the Free Software                       
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.    
+####################################################################################
+
+####################################################################################
+#  File hpdsampling.R
+#  This code is a distributed version of sample availabel in R gbm package.
+
+####################################################################################
+
+
+
+####################################################################################
+#                  Distributed sampling
+####################################################################################
+hpdsampling <- function(dfX,daY, Ns, npartition)
+{
+  #dfX: predictor variables in dframe/darray
+  #daY: vector of outcomes in darray (dvector/dfactor is not supported yet in2015-7-01
+  #npartition: now the sampled data have the same npartition as input. In the future, the input and output may have different partitions
+  #Ns : size of sub-chunks. #Ns <- 1*ceiling((nTrain/npartition)/npartition) # size of sampled sub-chunk
+  #function output: dfRX, daRY                 
+
+  ####################################################################################
+  # check function arguments
+   if(missing(dfX))
+	stop("'dfX' is a required argument")
+
+   if(missing(daY))
+	stop("'Y_train' is a required argument")
+
+   if(missing(Ns))
+	stop("'Ns' is a required argument")
+
+   if(missing(npartition))
+	stop("'npartition' is a required argument")
+
+   if(!is.dframe(dfX) && !is.darray(dfX))
+       stop("'dfX' must be a dframe or darray")
+
+   # for binary classification/regression, daY is a darray. For multinomial distribution (multi-classification), daY is a dframe (factor vecotr originally)
+   if(!is.dframe(daY) && !is.darray(daY))
+       stop("'daY' must be a dframe or darray")
+
+    if (npartition <= 0)
+       stop("'npartition' must be a positive number")
+
+    if (Ns <= 0)
+       stop("'Ns' must be a positive number")
+
+
+
+  ### distributed sampling of X_train, Y_train
+  p <- ncol(dfX) # p: number of variables
+  dfSX <- dframe(c(Ns*npartition,p), blocks=c(Ns,p))  
```

Comment:
Vishrut,

Could you please write a few lines of codes to implement what you suggest?

In our  grouping discussion, I remembered there is no direct worker-to-worker data transmission. So I designed distributed sampling like this. I cannot find a programming guide about how to transmit data between workers effectively. Where can I find the programming guide about what you mentioned in this email?

Thanks!

Winston



From: vishrutg [mailto:notifications@github.com]
Sent: Thursday, August 13, 2015 3:05 PM
To: vertica/DistributedR
Cc: Li, Winston
Subject: Re: [DistributedR] adding hpdegbm codes and man (#19)


In algorithms/HPdclassifier/R/hpdsampling.R<https://github.com/vertica/DistributedR/pull/19#discussion_r37012843>:

> +

> +   # for binary classification/regression, daY is a darray. For multinomial distribution (multi-classification), daY is a dframe (factor vecotr originally)

> +   if(!is.dframe(daY) && !is.darray(daY))

> +       stop("'daY' must be a dframe or darray")

> +

> +    if (npartition <= 0)

> +       stop("'npartition' must be a positive number")

> +

> +    if (Ns <= 0)

> +       stop("'Ns' must be a positive number")

> +

> +

> +

> +  ### distributed sampling of X_train, Y_train

> +  p <- ncol(dfX) # p: number of variables

> +  dfSX <- dframe(c(Ns*npartition,p), blocks=c(Ns,p))

When you are sampling and then putting all sampled values into a partition, you are forcing the data to first go to the master and then the worker (because you used getpartition). Instead you can directly send the data from all workers to 1 worker via splits(SX_train). By not providing an index it will send all partitions to the worker.

Also, You have created a for loop outside of the 2 foreach. It would be faster to collapse all of this into 2 foreach loops. You can use the enhanced foreach syntax to do this: splits(A,list(1,2,3)) will put partitions 1,2,3 of dobject A into the same executor.

—
Reply to this email directly or view it on GitHub<https://github.com/vertica/DistributedR/pull/19/files#r37012843>.